### ISSUE TITLE
Deduplicate "Making Change"

### ISSUE THREAD

**Issue by hobovsky** [codewars/content-issues#93]
Follow-up on #92.

Note on languages: If not specified otherwise, all kata are available in JavaScript, Ruby, and Python additionally to explicitly listed languages.

1. [Loose Change](https://www.codewars.com/kata/5571f712ddf00b54420000ee)
- 6 kyu.
- Return value is dictionary, does not seem to have performance reqs.
- Satisfaction 92%, ~4k solutions.
- 2 pending issues (1 about being a duplicate).
- 4 languages: 3 of them overlap with other kata. No pending translations. Available in Java.
- Published Jun 2015.

2. [Making Change](https://www.codewars.com/kata/5365bb5d5d0266cd010009be)
- 6 kyu
- Return value is a dictionary, does not seem to have performance reqs
- Satisfaction 95%, ~1k solutions
- 2 pending issues
- 5 languages + 1 pending translation. Available in Elixir, Haskell, Julia
- Published May 2014 

3. [You Got Change?](https://www.codewars.com/kata/5966f6343c0702d1dc00004c)
- 7 kyu
- Return value is an array, a bit more performance heavy but still O(n) solutions pass easily.
- Satisfaction 92%, ~1.5k solutions.
- 2 pending issues, one related to being a duplicate.
- 5 languages + 2 pending translations. Available in C, C#, COBOL, Julia
- Published Jul 2017

4. [ATM](https://www.codewars.com/kata/5635e7cb49adc7b54500001c)
- 7 kyu
- Return value is only amount of notes, without a breakdown into denominations. No performance reqs.
- Satisfaction 93%, ~4.5k solutions.
- No pending issues, but would require updates of some frameworks.
- 8 languages + 2 pending translations. Available in C, CoffeeScript, F#, Java, Kotlin, Cobol, Rust. 
- Published Nov 2015

5. [Change Machine](https://www.codewars.com/kata/537ece6953efd95e9e000516)
- 6 kyu
- Return value is a dictionary. Reported to have performance requirements.
- Satisfaction 84%, 357 solutions.
- 5 pending issues, one related to being a duplicate.
- Available _only_ in Ruby and CoffeeScript. I haven't solved it so I cannot check pending translations, but none are suggested in discourse.
- Published May 2014.

Added:

6. [Minimum dollar bill's count](https://www.codewars.com/kata/58e4d3530e1018e155000058)
- 6 kyu
- Return count of bills. Contains large-ish inputs, but still O(n) solutions pass without problem.
- Satisfaction 90%, 123 solutions.
- One pending issue about outdated testing framework.
- Available only in Javascript.
- Published Apr 2017.

Added:

7. [No one likes carrying change.](https://www.codewars.com/kata/58dfc0d8ca6895cb4e00005c/)
- 6 kyu
- Return count of coins. Contains performance tests with large inputs, O(n) solutions don't pass. However, rules of the task are not clear and contradictory to itself, because it asks about both minimum number of coins and a greedy solution at the same time. 
- Satisfaction 83%, ~400 solutions.
- Two pending issues about unclear specs.
- Two languages and two pending translations, mostly overlapping with remaining kata listed here.
- Published Apr 2017.

Listed kata are very similar, the most important difference seems to be available languages.  
If languages affect your vote, I am willing to transfer all language versions which would get removed by deduplication before they get retired (I might need some help with COBOL, but doh, maybe I finally learn it :) )

**Comment by hobovsky** [comment]
I solved all kata except the last one in Python. I used O(1) solution in each of them. All of them differ only in terms of return format.

Kata 4. has the most languages, but also the worst return format. It's my least favorite candidate to keep, but probably the one which can be kept with the least effort.  
From 1., 2., 3., I have no good candidate to keep because all of them are almost the same. I think I'd keep 3. because the other two have languages I know how to transfer :)  
I don't think we need 5. because it does not bring anything new, really. It's said to have performance requirements, but I was not able to verify them.

My vote would be:
- Retire 1. and 2., keep 3.
- I don;t know what to do with 4. It's similar to others but has many languages which would take some effort to transfer. If you think we really don't need it, we can retire it.
- I am not sure what to do with 5. It's the only one with reported performance requirements, but I am fine both with keeping it and retiring it.
- Added: 6. is identical as 4., so I vote to retire it.
- Added: 7. would be nice because it has performance reqs and _potentially_ asks about a different algorithm, but execution is simply bad. Additionally, no matter if the required algorithm would be greedy or not, it's going to end up as a duplicate. I'd just vote to retire.

**Comment by ejini6969** [comment]
In 5, there's no random test, only a test with very large input, so no performance req there! 

BTW, I found another one yesterday -> [Minimum dollar bill's count](https://www.codewars.com/kata/58e4d3530e1018e155000058/discuss)
* 6kyu 
* JS only
* 1 pending issue
* No pending translations
* Published April 5 2017
* Similar to 4 (ATM) except it always has a valid answer (no -1)

My opinion: 1, 2, 3 & 5 are similar, 4 & above kata are similar but the technique used from both groups seem to not vary too much to make it a non-duplicate. 

Nonetheless, I voted to keep 4 (ATM) (as it considers invalid cases which is better compared to above kata IMO + having highest completion among them all)

For the other grp, I voted to keep 2 since it has the most languages && it's the earliest one to be published)
* 1 is too troublesome to maintain, needs framework update && considers negative money count, which is ridiculous
* 3 is ranked wrongly IMO and needs to differentiate with 4 (ATM) 
* 5 has many issues && only 2 not-so-popular languages left (high maintenance cost also)

**Comment by akar-0** [comment]
Keep 3 only. The others are duplicate, sometimes with input validation, 5 has bad tests (no sample tests, no random tests). I'm voluntary to participate to the transfer of missing languages, I think it should not be that hard.

**Comment by hobovsky** [comment]
> BTW, I found another one yesterday -> [Minimum dollar bill's count](https://www.codewars.com/kata/58e4d3530e1018e155000058/discuss)

I added it to the list.

**Comment by CiprianAmza** [comment]
All of them ask pretty much the same. I propose to keep (2) Making Change (since for a beginner the use of a map/dictionary/hash can be delicate and it's an useful exercise) and (5) ATM (which only asks for the total number and has the most translations).

**Comment by EloiseRosen** [comment]
The tests for 5 suck, it doesn't even have a sample test in either of its languages, and no random tests.
I vote for keeping just one of these and retiring the rest, with no strong opinion on which to keep besides "not 5".

**Comment by rowcased** [comment]
This one should be added:
<a href= "https://www.codewars.com/kata/586b1b91c66d181c2c00016f/python">Vending Machine Problems [#1]</a>

**Comment by ejini6969** [comment]
> This one should be added: [Vending Machine Problems [#1]](https://www.codewars.com/kata/586b1b91c66d181c2c00016f/python)

No, this one involves a higher level problem ~~See #39 -> Minimum change !

**Comment by akar-0** [comment]
[No one likes carrying change.](https://www.codewars.com/kata/58dfc0d8ca6895cb4e00005c) is another duplicate...

**Comment by ejini6969** [comment]
> [No one likes carrying change.](https://www.codewars.com/kata/58dfc0d8ca6895cb4e00005c) is another duplicate...

Indeed, I voted to retire this one too ~~ due to having least amount of languages & a lot of pending issues

**Comment by hobovsky** [comment]
> [No one likes carrying change.](https://www.codewars.com/kata/58dfc0d8ca6895cb4e00005c) is another duplicate...

Added to the list.

**Comment by fcr--** [comment]
Retire: 1, 2, 4, 5
Keep: 3, 7

6... well, a naïve greedy algorithm could result in a non-optimal partition.  For example let's look at: 400, [201, 100, 1]; where the naïve approach would result in 201×1+100×1+1×99 (101 bills), it could be split optimally with 100×4 (4 bills).  However the current test cases are designed so that the greedy algorithm works.  So for this one I think that we should either:
* make it general enough so that it could work with random note values (but this would invalidate almost all solutions, and it would become a way more difficult kata ~4kyu I'd say... I had to do this: https://www.codewars.com/kata/reviews/58e4eb18d8fb2ee3990013a1/groups/63013c29efce070001e784b5), or
* retire it if the notes are left at 1,10,20,50,100; since it would be a duplicate of 7.

**Comment by monadius** [comment]
Keep 3 and 7.

**Comment by Kacarott** [comment]
I agree with keeping 3 and 7

**Comment by 4500zenja1** [comment]
By the look of the katas and their requirements, I'd split them into two groups, "Return the amount of *separate coins*" (1, 2, 3, 5) and "Return the overall amount of coins/notes" (4, 6, 7).

1. Story- and popularity-wise, 1 should be kept as more concise and user-friendly kata, BUT time-wise and sanity-wise, 2 has been pulished earlier + doesn't deal with negative/float inputs. 3 has array output, has more known langs like C/C# and decent amount of solutions but published much later than others (2017 vs 2014). 5 is already a lost case since the lowest solution and lang count, lots of issues, etc.

For that group, 1 and 5 are retire, 2 and 3 are keep (among them, though, 3 seems to be better choice).

2. For a simple approach, keep 4 and retire 6 as its duplicate; for nuanced one with different sets of coins, *maybe* keep 7, but not sure about "minimum number" and "greedy" since they're kinda contradictory... 

Overall, keep 3 and 4, maybe with 2 (as the oldest) or 7 (with not fixed coins BUT only if the issues are addresses), the rest are safe to retire.
